### ISSUE TITLE
[Proposal] Expose ikvm.lang.Property

### ISSUE THREAD

**Issue by wasabii** [ikvm-revived/ikvm#247]
This originates from my conversation with @joniles at #244.

Jon runs the MPXJ project, which is an open source Java library for Microsoft Project (and other formats). He has, for 10 years now, published a .NET version of his project. Part of his project is a produced .NET version, relying on IKVM. In using IKVM he relies on features that are available when manipulating ikvmc by hand such as map.xml, but are not available, and should not be available, when using MavenReference.

One of the biggest portions of his map.xml usage is the creation of CLR properties that correspond to his Java getters/setters. He does this by generating a map.xml file using some custom Java code that enumerates his classes.

IKVM has, for a long time now, had a hidden annotation at ikvm.lang.Property that allows you to decorate a Java field and specify the getter/setter method names. During compilation this generates a CLR property. Same as MPXJ does, but built in.

An annotation like this could fill a missing gap when using MavenReference. Our biggest limitation with MavenReference is that we need to consistently generate assemblies from Maven artifacts. People build ABI/API dependencies on MavenReferences. And they publish assemblies with these dependencies to NuGet. Thus, we can't allow arbitrary projects to alter how they generate assemblies. There's no way to harmonize that in a dependency graph.

However, having a feature for project owners (Jon) to annotate his Java classes in Maven with information that IKVM can consume is fine. The information is centralized with the Maven artifact, and thus can be used by any consumer of that artifact on IKVM. And each consumer can always generate the same assembly from that information.

So this proposal is to make ikvm.lang.Property publically available for Java authors to use within their project. Their real Java code will make use of this property. It will be ignored by standard JVMs.

For authors to use this, it has to be available. Right now the annotation is compiled within IKVM.Java, which isn't itself accessible from Java. We don't support generating stubs for IKVM.Java. And it's not like we distribute this in a JAR.

I propose we move at least ikvm.lang.Property out of IKVM.Java, into it's own Maven project. A real Maven project: not MSBuild. And that this resulting JAR file is published by us to Maven as part of our release cycle. Java authors can then consume it in their own Maven projects, as a compile dependency. Allowing them to annotate their types. When those JAR files are eventually published, and then eventually consumed by MavenReference, this metadata can cause the properties to be generated within IKVM.

**Comment by wasabii** [comment]
To note: this proposal kind of overlaps with, but also has nothing to do with, the Java Bean "Standard". CLR properties aren't seen by Java code. Any bean-stuff going on will continue to see the getFoo and setFoo methods.

We could do something magic like auto infer properties from get/set methods. But this is really a terrible idea.

**Comment by joniles** [comment]
Following on from the conversation in https://github.com/ikvm-revived/ikvm/issues/244 I've had a look at the current `ikvm.lang.Property` implementation. This will work fine for simple instance variables where there is a one-to-one mapping between the instance variable and getter/setter methods.

My use case is slightly different as most of the time I won't have an instance variable to which this property can be applied.

My code will typically look like this:

```java
class Task
{
  public void setManualDuration(Duration dur)
   {
      set(TaskField.MANUAL_DURATION, dur);
   }

   public Duration getManualDuration()
   {
      return (Duration) get(TaskField.MANUAL_DURATION);
   }
}
```

I would need an annotation which works something like this at the class level:

```java
@ikvm.lang.Property(name = "ManualDuration", get = "getManualDuration", set = "setManualDuration")
class Task
{
  public void setManualDuration(Duration dur)
   {
      set(TaskField.MANUAL_DURATION, dur);
   }

   public Duration getManualDuration()
   {
      return (Duration) get(TaskField.MANUAL_DURATION);
   }
}
```

or

```java
class Task
{
  @ikvm.lang.SetProperty("ManualDuration")
  public void setManualDuration(Duration dur)
   {
      set(TaskField.MANUAL_DURATION, dur);
   }

   @ikvm.lang.GetProperty("ManualDuration")
   public Duration getManualDuration()
   {
      return (Duration) get(TaskField.MANUAL_DURATION);
   }
}
```

As we have a well understood Java convention for naming getter and setter methods, we also have the option to simplify the annotation, for example:

```java
@ikvm.lang.Property(name = "ManualDuration")
class Task
{
 // ...
}
```

or even something like this so all properties can be defined in one place:

```java
@ikvm.lang.Properties(names ={
  "Duration",
  "Start",
  "ManualDuration"})
class Task
{
 // ...
}
```

**Comment by wasabii** [comment]
>As we have a well understood Java convention for naming getter and setter methods, we also have the option to simplify the annotation, for example:

This is sort of the one thing I want to avoid, as the Java Bean standard isn't really that much of a standard.

I do think ikvm.lang.Property should be more flexible.

**Comment by joniles** [comment]
> This is sort of the one thing I want to avoid

I agree with your desire to keep things explicit. My intention would be that the primary annotation should always look like this:

```java
@ikvm.lang.Property(name = "ManualDuration", get = "getManualDuration", set = "setManualDuration")
```

with the option of shortening to 

```java
@ikvm.lang.Property(name = "ManualDuration")
```

_if_ your getter and setter method names just prefix the property name with `get` and `set`.

With regard to ABI, both approaches give you the same result: the property is named explicitly and it should be clear to you as a Java developer than you are making a breaking change when you edit the name attribute in the annotation which will affect the property for .Net users.

Whichever form you choose, IKVM should fail with an error if the method names are not recognised regardless of whether you have supplied them explicitly, or IKVM has inferred them from the property name. A suitably detailed error message should be able to highlight the problem to the Java maintainer.
